File: src/sdk/llmwatch.ts
// src/sdk/llmwatch.ts

export interface LLMWatchConfig {
  apiKey: string;
  endpoint?: string;
  flushIntervalMs?: number;
  maxBatchSize?: number;
}

export interface TrackEvent {
  idempotency_key: string;
  request_id?: string;
  prompt_hash?: string;
  customer_identifier?: string;
  provider: string;
  model: string;
  prompt_tokens?: number;
  completion_tokens?: number;
  cost_usd?: number;
  latency_ms?: number;
  error_message?: string;
  is_cached?: boolean;
  request_payload?: any;
  response_payload?: any;
  trace_id?: string;
  parent_span_id?: string;
  span_type?: 'llm' | 'tool' | 'chain' | 'agent';
  span_name?: string;
  execution_order?: number;
  state_snapshot?: any;
  reasoning_text?: string;
  duration_breakdown?: {
    queue_ms?: number;
    model_ms?: number;
    tool_ms?: number;
    overhead_ms?: number;
    [key: string]: any;
  };
  [key: string]: any;
}

// Global-friendly UUID generator fallback
function generateUUID(): string {
  let cryptoObj = typeof globalThis !== 'undefined' ? (globalThis as any).crypto : null;
  if (!cryptoObj && typeof window !== 'undefined') {
    cryptoObj = window.crypto;
  }
  if (cryptoObj && typeof cryptoObj.randomUUID === 'function') {
    return cryptoObj.randomUUID();
  }
  // Fallback UUID v4 generator
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = Math.random() * 16 | 0;
    const v = c === 'x' ? r : (r & 0x3 | 0x8);
    return v.toString(16);
  });
}

// Simple prompt hashing to generate prompt_hash if not provided
function simpleHash(text: string): string {
  if (!text) return '';
  let hash = 0;
  for (let i = 0; i < text.length; i++) {
    const char = text.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash |= 0; // Convert to 32bit integer
  }
  return Math.abs(hash).toString(16);
}

export class Span {
  public id: string;
  public traceId: string;
  public parentSpanId: string | null;
  public name: string;
  public type: 'llm' | 'tool' | 'chain' | 'agent';
  private sdk: LLMWatch;
  private startTime: number;
  private executionCounter: { count: number };
  public executionOrder: number;
  private stateSnapshot?: any;
  private reasoningText?: string;
  private durationBreakdownVal?: any;

  constructor(
    sdk: LLMWatch,
    name: string,
    type: 'llm' | 'tool' | 'chain' | 'agent' = 'llm',
    traceId?: string,
    parentSpanId: string | null = null,
    executionCounter?: { count: number }
  ) {
    this.sdk = sdk;
    this.name = name;
    this.type = type;
    this.id = generateUUID();
    this.traceId = traceId || generateUUID();
    this.parentSpanId = parentSpanId;
    this.startTime = Date.now();
    this.executionCounter = executionCounter || { count: 0 };
    this.executionOrder = this.executionCounter.count++;
  }

  public span(name: string, options?: { spanType?: 'llm' | 'tool' | 'chain' | 'agent' }): Span {
    return new Span(this.sdk, name, options?.spanType || 'llm', this.traceId, this.id, this.executionCounter);
  }

  public captureState(state: any): Span {
    this.stateSnapshot = state;
    return this;
  }

  public reasoning(text: string): Span {
    this.reasoningText = text;
    return this;
  }

  public durationBreakdown(breakdown: any): Span {
    this.durationBreakdownVal = breakdown;
    return this;
  }

  public track(data: Partial<TrackEvent>) {
    const event: TrackEvent = {
      idempotency_key: this.id,
      trace_id: this.traceId,
      parent_span_id: this.parentSpanId || undefined,
      span_type: this.type,
      span_name: this.name,
      provider: data.provider || 'unknown',
      model: data.model || 'unknown',
      execution_order: this.executionOrder,
      state_snapshot: this.stateSnapshot,
      reasoning_text: this.reasoningText,
      duration_breakdown: this.durationBreakdownVal,
      ...data
    };
    this.sdk.track(event);
  }

  public end(data?: Partial<TrackEvent>) {
    const latency_ms = Date.now() - this.startTime;
    this.track({ latency_ms, ...data });
  }
}

export class LLMWatch {
  private apiKey: string;
  private endpoint: string;
  private buffer: TrackEvent[] = [];
  private flushIntervalMs: number;
  private maxBatchSize: number;
  private timer: ReturnType<typeof setInterval> | null = null;

  constructor(config: LLMWatchConfig) {
    this.apiKey = config.apiKey;
    this.endpoint = config.endpoint || 'http://localhost:3000/api/public/v1/events';
    this.flushIntervalMs = config.flushIntervalMs || 5000;
    this.maxBatchSize = config.maxBatchSize || 100;

    // Start timer for flush intervals
    this.timer = setInterval(() => this.flush(), this.flushIntervalMs);
    
    // Graceful draining on process exit (Node environment)
    if (typeof process !== 'undefined' && typeof process.on === 'function') {
      process.on('SIGINT', () => this.close());
      process.on('SIGTERM', () => this.close());
    }
  }

  public trace(name: string, options?: { spanType?: 'llm' | 'tool' | 'chain' | 'agent'; traceId?: string }): Span {
    return new Span(this, name, options?.spanType || 'agent', options?.traceId || generateUUID(), null);
  }

  public track(event: TrackEvent) {
    // Fill in default uuid if missing
    if (!event.idempotency_key) {
      event.idempotency_key = generateUUID();
    }
    this.buffer.push(event);
    
    if (this.buffer.length >= this.maxBatchSize) {
      this.flush();
    }
  }

  public async flush() {
    if (this.buffer.length === 0) return;
    
    const batch = this.buffer.splice(0, this.maxBatchSize);
    let attempts = 0;
    
    while (attempts < 3) {
      try {
        const res = await fetch(this.endpoint, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${this.apiKey}`
          },
          body: JSON.stringify(batch)
        });
        
        // Don't retry client-side errors (400-499)
        if (res.ok || (res.status >= 400 && res.status < 500)) {
          break;
        }
        
      } catch (err) {
        // Network error - continue to retry
      }
      
      attempts++;
      if (attempts < 3) {
        // Exponential backoff retry: 1s, 2s, capped at 8s
        const delay = Math.min(1000 * Math.pow(2, attempts - 1), 8000);
        await new Promise(r => setTimeout(r, delay));
      }
    }
  }

  public async close() {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    await this.flush();
  }

  // Deterministic hashing for prompt parameters
  private async generateCacheKey(params: any): Promise<string> {
    // Sort keys recursively/shallowly so {"model":"x", "temp":1} hashes the same as {"temp":1, "model":"x"}
    const sortedParams = Object.keys(params).sort().reduce((acc: any, key) => {
      acc[key] = params[key];
      return acc;
    }, {});
    
    const payloadStr = JSON.stringify(sortedParams);
    const msgUint8 = new TextEncoder().encode(payloadStr);
    
    let cryptoObj = typeof globalThis !== 'undefined' ? (globalThis as any).crypto : null;
    if (!cryptoObj && typeof window !== 'undefined') {
      cryptoObj = window.crypto;
    }
    
    if (cryptoObj && cryptoObj.subtle) {
      const hashBuffer = await cryptoObj.subtle.digest('SHA-256', msgUint8);
      const hashArray = Array.from(new Uint8Array(hashBuffer));
      return hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
    } else {
      // Fallback for older Node.js runtimes
      try {
        const nodeCrypto = require('crypto');
        return nodeCrypto.createHash('sha256').update(payloadStr).digest('hex');
      } catch (err) {
        // Simple fallback hash if node crypto is not available
        return simpleHash(payloadStr);
      }
    }
  }

  // Auto-instrumentation wrapper for OpenAI
  public wrapOpenAI(openaiClient: any, options?: { enableCache?: boolean; cacheTtl?: number }) {
    const originalCreate = openaiClient.chat.completions.create.bind(openaiClient.chat.completions);
    const cacheEndpoint = this.endpoint.replace('/events', '/cache');
    
    openaiClient.chat.completions.create = async (params: any, reqOptions?: any) => {
      const start = Date.now();
      const idempotency_key = generateUUID();
      
      // Extract prompt text for prompt hash
      let promptText = '';
      if (params.messages && Array.isArray(params.messages)) {
        promptText = JSON.stringify(params.messages);
      } else {
        promptText = JSON.stringify(params);
      }
      const prompt_hash = simpleHash(promptText);
      
      // 1. Check Cache if enabled (and not streaming)
      let cacheKey = '';
      if (options?.enableCache && !params.stream) {
        cacheKey = await this.generateCacheKey(params);
        try {
          const cacheRes = await fetch(`${cacheEndpoint}?key=${cacheKey}`, {
            headers: { 'Authorization': `Bearer ${this.apiKey}` }
          });
          
          if (cacheRes.ok) {
            const cacheData = await cacheRes.json();
            if (cacheData.hit) {
              // Track as a 0-cost, 50ms latency cached hit!
              this.track({
                idempotency_key,
                provider: 'openai',
                model: params.model,
                prompt_hash,
                latency_ms: Date.now() - start,
                is_cached: true,
                cost_usd: 0, // Cache hits are free
                request_payload: params,
                response_payload: cacheData.data
              });
              
              return cacheData.data; // Return identical cached payload
            }
          }
        } catch (e) {
          console.warn('[LLMWatch] Cache fetch failed, falling back to network', e);
        }
      }
      
      // 2. Call OpenAI on Cache Miss
      try {
        const response = await originalCreate(params, reqOptions);
        
        // Track non-stream completions
        if (!params.stream) {
          this.track({
            idempotency_key,
            provider: 'openai',
            model: response.model || params.model,
            prompt_hash,
            prompt_tokens: response.usage?.prompt_tokens,
            completion_tokens: response.usage?.completion_tokens,
            latency_ms: Date.now() - start,
            is_cached: false,
            request_payload: params,
            response_payload: response
          });

          // 3. Asynchronously write back to cache
          if (options?.enableCache && cacheKey) {
            fetch(cacheEndpoint, {
              method: 'POST',
              headers: { 
                'Content-Type': 'application/json', 
                'Authorization': `Bearer ${this.apiKey}` 
              },
              body: JSON.stringify({
                cacheKey,
                provider: 'openai',
                model: params.model,
                payload: response,
                ttlSeconds: options.cacheTtl || 86400
              })
            }).catch((err) => {
              console.warn('[LLMWatch] Async cache write-back failed', err);
            });
          }
        }
        return response;
      } catch (error: any) {
        // Track API execution failures
        this.track({
          idempotency_key,
          provider: 'openai',
          model: params.model,
          prompt_hash,
          latency_ms: Date.now() - start,
          error_message: error.message || 'OpenAI API execution error',
          request_payload: params
        });
        throw error;
      }
    };
    
    return openaiClient;
  }
}

File: src/sdk/llmwatch.py
# src/sdk/llmwatch.py
import json
import time
import uuid
import hashlib
import threading
import queue
import atexit
import urllib.request
import urllib.error

class Span:
    def __init__(self, sdk, name, span_type="llm", trace_id=None, parent_span_id=None, execution_counter=None):
        self.sdk = sdk
        self.name = name
        self.type = span_type
        self.id = str(uuid.uuid4())
        self.trace_id = trace_id or str(uuid.uuid4())
        self.parent_span_id = parent_span_id
        self.start_time = time.time()
        self.ended = False
        self._state_snapshot = None
        self._reasoning_text = None
        self._duration_breakdown = None
        self._token_breakdown = None
        self._context_window_used = None
        self._context_window_max = None

        if execution_counter is None:
            self.execution_counter = [0]
        else:
            self.execution_counter = execution_counter
        
        self.execution_order = self.execution_counter[0]
        self.execution_counter[0] += 1

    def span(self, name, span_type="llm"):
        return Span(self.sdk, name, span_type=span_type, trace_id=self.trace_id, parent_span_id=self.id, execution_counter=self.execution_counter)

    def capture_state(self, state_dict):
        self._state_snapshot = state_dict
        return self

    def reasoning(self, text):
        self._reasoning_text = text
        return self

    def duration_breakdown(self, breakdown):
        self._duration_breakdown = breakdown
        return self

    def token_breakdown(self, breakdown):
        self._token_breakdown = breakdown
        return self

    def context_window(self, used, max_tokens=None):
        self._context_window_used = used
        self._context_window_max = max_tokens
        return self

    def track(self, **data):
        event = {
            "idempotency_key": self.id,
            "trace_id": self.trace_id,
            "parent_span_id": self.parent_span_id,
            "span_type": self.type,
            "span_name": self.name,
            "provider": data.get("provider", "unknown"),
            "model": data.get("model", "unknown"),
            "execution_order": self.execution_order,
        }
        if self._state_snapshot is not None:
            event["state_snapshot"] = self._state_snapshot
        if self._reasoning_text is not None:
            event["reasoning_text"] = self._reasoning_text
        if self._duration_breakdown is not None:
            event["duration_breakdown"] = self._duration_breakdown
        if self._token_breakdown is not None:
            event["token_breakdown"] = self._token_breakdown
        if self._context_window_used is not None:
            event["context_window_used"] = self._context_window_used
        if self._context_window_max is not None:
            event["context_window_max"] = self._context_window_max

        event.update(data)
        self.sdk.track(event)

    def end(self, **data):
        if not self.ended:
            self.ended = True
            latency_ms = int((time.time() - self.start_time) * 1000)
            data["latency_ms"] = latency_ms
            self.track(**data)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        error_message = None
        if exc_type is not None:
            error_message = f"{exc_type.__name__}: {str(exc_val)}"
        
        if error_message:
            self.end(error_message=error_message)
        else:
            self.end()
        return False

class LLMWatch:
    def __init__(self, api_key, endpoint=None, flush_interval_ms=5000, max_batch_size=100):
        self.api_key = api_key
        self.endpoint = endpoint or "http://localhost:3000/api/public/v1/events"
        self.flush_interval_seconds = flush_interval_ms / 1000.0
        self.max_batch_size = max_batch_size
        
        self.queue = queue.Queue()
        self.running = True
        
        # Background worker thread for non-blocking flushes
        self.worker_thread = threading.Thread(target=self._worker, daemon=True)
        self.worker_thread.start()
        
        # Ensure queue drains cleanly on process shutdown
        atexit.register(self.close)

    def trace(self, name, span_type="agent", trace_id=None):
        """Create a root tracking trace (returns a Span)."""
        return Span(self, name, span_type=span_type, trace_id=trace_id, parent_span_id=None)

    def track(self, event):
        """Track custom telemetry events."""
        if not isinstance(event, dict):
            return
            
        if "idempotency_key" not in event:
            event["idempotency_key"] = str(uuid.uuid4())
            
        self.queue.put(event)
        
        if self.queue.qsize() >= self.max_batch_size:
            # Trigger quick background flush (non-blocking thread notification)
            pass

    def _worker(self):
        last_flush_time = time.time()
        while self.running:
            try:
                # Wait for items or flush timeout
                time_elapsed = time.time() - last_flush_time
                wait_timeout = max(0.1, self.flush_interval_seconds - time_elapsed)
                
                try:
                    event = self.queue.get(timeout=wait_timeout)
                    self.queue.task_done()
                except queue.Empty:
                    event = None

                # Flush if max batch size reached, or interval timer expired
                if self.queue.qsize() >= self.max_batch_size or (time.time() - last_flush_time >= self.flush_interval_seconds):
                    self.flush()
                    last_flush_time = time.time()
            except Exception as e:
                print(f"[LLMWatch] Internal worker exception: {e}")

    def flush(self):
        """Drains the buffer and posts events in batches with exponential backoff retries."""
        batch = []
        while not self.queue.empty() and len(batch) < self.max_batch_size:
            try:
                batch.append(self.queue.get_nowait())
                self.queue.task_done()
            except queue.Empty:
                break

        if not batch:
            return

        attempts = 0
        while attempts < 3:
            try:
                req = urllib.request.Request(
                    self.endpoint,
                    data=json.dumps(batch).encode("utf-8"),
                    headers={
                        "Content-Type": "application/json",
                        "Authorization": f"Bearer {self.api_key}"
                    },
                    method="POST"
                )
                
                with urllib.request.urlopen(req, timeout=8) as response:
                    if response.status == 202 or (400 <= response.status < 500):
                        break # Success or client-side failure (don't retry 4xx errors)
            except urllib.error.HTTPError as e:
                if 400 <= e.code < 500:
                    break # Client error - do not retry
            except Exception:
                pass # Network error - retry
                
            attempts += 1
            if attempts < 3:
                delay = min(1.0 * (2 ** (attempts - 1)), 8.0)
                time.sleep(delay)

    def close(self):
        """Flushes remaining items and stops the background thread gracefully."""
        if self.running:
            self.running = False
            # Drain whatever is left
            limit = 100
            while not self.queue.empty() and limit > 0:
                self.flush()
                limit -= 1

    def _generate_cache_key(self, params):
        # Sort prompt dictionary keys recursively for deterministic hash signature
        payload_str = json.dumps(params, sort_keys=True)
        return hashlib.sha256(payload_str.encode("utf-8")).hexdigest()

    def wrap_openai(self, openai_client, enable_cache=False, cache_ttl=86400):
        """Monkeypatches the OpenAI client completions call to intercept cache and track telemetry."""
        original_create = openai_client.chat.completions.create
        cache_endpoint = self.endpoint.replace("/events", "/cache")

        def wrapped_create(*args, **kwargs):
            start_time = time.time()
            idempotency_key = str(uuid.uuid4())
            
            # Extract prompt messages
            messages = kwargs.get("messages", [])
            prompt_hash = hashlib.sha256(json.dumps(messages).encode("utf-8")).hexdigest()
            model = kwargs.get("model", "unknown")
            is_stream = kwargs.get("stream", False)
            
            cache_key = ""
            if enable_cache and not is_stream:
                cache_key = self._generate_cache_key(kwargs)
                try:
                    # Query cache hit
                    req_url = f"{cache_endpoint}?key={cache_key}"
                    req = urllib.request.Request(
                        req_url,
                        headers={"Authorization": f"Bearer {self.api_key}"},
                        method="GET"
                    )
                    with urllib.request.urlopen(req, timeout=3) as res:
                        cache_data = json.loads(res.read().decode("utf-8"))
                        if cache_data.get("hit"):
                            # Log zero-cost cache hit telemetry
                            latency_ms = int((time.time() - start_time) * 1000)
                            self.track({
                                "idempotency_key": idempotency_key,
                                "provider": "openai",
                                "model": model,
                                "prompt_hash": prompt_hash,
                                "latency_ms": latency_ms,
                                "is_cached": True,
                                "cost_usd": 0.0,
                                "request_payload": kwargs,
                                "response_payload": cache_data["data"]
                            })
                            
                            # Reconstruct mock class mimicking OpenAI response
                            return _mock_openai_response(cache_data["data"])
                except Exception as e:
                    print(f"[LLMWatch] Caching lookup bypassed due to error: {e}")

            # Execute real OpenAI call on cache miss
            try:
                response = original_create(*args, **kwargs)
                
                if not is_stream:
                    latency_ms = int((time.time() - start_time) * 1000)
                    # Convert response to dictionary representation
                    res_dict = _openai_response_to_dict(response)
                    
                    self.track({
                        "idempotency_key": idempotency_key,
                        "provider": "openai",
                        "model": res_dict.get("model", model),
                        "prompt_hash": prompt_hash,
                        "prompt_tokens": res_dict.get("usage", {}).get("prompt_tokens", 0),
                        "completion_tokens": res_dict.get("usage", {}).get("completion_tokens", 0),
                        "latency_ms": latency_ms,
                        "is_cached": False,
                        "request_payload": kwargs,
                        "response_payload": res_dict
                    })

                    # Write response back to cache asynchronously
                    if enable_cache and cache_key:
                        def async_write():
                            try:
                                post_req = urllib.request.Request(
                                    cache_endpoint,
                                    data=json.dumps({
                                        "cacheKey": cache_key,
                                        "provider": "openai",
                                        "model": model,
                                        "payload": res_dict,
                                        "ttlSeconds": cache_ttl
                                    }).encode("utf-8"),
                                    headers={
                                        "Content-Type": "application/json",
                                        "Authorization": f"Bearer {self.api_key}"
                                    },
                                    method="POST"
                                )
                                with urllib.request.urlopen(post_req, timeout=3) as write_res:
                                    pass
                            except Exception:
                                pass # Silently ignore cache write failures
                                
                        threading.Thread(target=async_write, daemon=True).start()

                return response
            except Exception as error:
                latency_ms = int((time.time() - start_time) * 1000)
                self.track({
                    "idempotency_key": idempotency_key,
                    "provider": "openai",
                    "model": model,
                    "prompt_hash": prompt_hash,
                    "latency_ms": latency_ms,
                    "error_message": getattr(error, "message", str(error)),
                    "request_payload": kwargs
                })
                raise error

        openai_client.chat.completions.create = wrapped_create
        return openai_client

# Helper functions to convert between dictionary models and OpenAI class definitions
def _openai_response_to_dict(response):
    if hasattr(response, "model_dump"):
        return response.model_dump()
    elif hasattr(response, "__dict__"):
        # Fallback for dict structures
        return json.loads(json.dumps(response, default=lambda o: o.__dict__))
    return response

def _mock_openai_response(data):
    # Simple recursive object wrapper matching dot-notation expectations of OpenAI SDK client response
    class OpenAIObject:
        def __init__(self, d):
            for k, v in d.items():
                if isinstance(v, dict):
                    setattr(self, k, OpenAIObject(v))
                elif isinstance(v, list):
                    setattr(self, k, [OpenAIObject(i) if isinstance(i, dict) else i for i in v])
                else:
                    setattr(self, k, v)
        
        def model_dump(self):
            return data
            
    return OpenAIObject(data)
